File: java/org.financer.client.javafx/src/main/java/org/financer/client/javafx/main/overview/OverviewController.java
package org.financer.client.javafx.main.overview;

import com.jfoenix.controls.JFXComboBox;
import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.geometry.HPos;
import javafx.scene.chart.PieChart;
import javafx.scene.chart.XYChart;
import javafx.scene.control.Hyperlink;
import javafx.scene.control.Label;
import javafx.scene.control.Tooltip;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import org.financer.client.domain.api.RestApi;
import org.financer.client.domain.api.RestApiImpl;
import org.financer.client.domain.model.category.Category;
import org.financer.client.domain.model.category.CategoryRoot;
import org.financer.client.domain.model.transaction.FixedTransaction;
import org.financer.client.domain.model.transaction.FixedTransactionAmount;
import org.financer.client.domain.model.transaction.Transaction;
import org.financer.client.domain.model.transaction.VariableTransaction;
import org.financer.client.domain.model.user.User;
import org.financer.client.format.I18N;
import org.financer.client.javafx.components.charts.DonutChart;
import org.financer.client.javafx.components.charts.SmoothedChart;
import org.financer.client.javafx.format.JavaFXFormatter;
import org.financer.client.javafx.local.LocalStorageImpl;
import org.financer.client.javafx.main.FinancerController;
import org.financer.client.javafx.main.transactions.TransactionAmountDialog;
import org.financer.shared.domain.model.value.objects.Amount;
import org.financer.shared.domain.model.value.objects.CategoryClass;
import org.financer.shared.domain.model.value.objects.TimeRange;
import org.financer.shared.domain.model.value.objects.ValueDate;
import org.financer.util.collections.TreeUtil;

import java.net.URL;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.ResourceBundle;
import java.util.concurrent.atomic.AtomicInteger;

public class OverviewController implements Initializable {
    @FXML
    public GridPane lastTransactionsGridPane;
    @FXML
    public GridPane balanceGridPane;
    @FXML
    public GridPane upcomingFixedTransactionGridPane;
    @FXML
    public Label balanceChangeLabel;
    @FXML
    public Label balanceLabel;
    @FXML
    public Label variableExpensesLabel;
    @FXML
    public Label variableExpensesChangeLabel;
    @FXML
    public Label numberOfTransactionsChangeLabel;
    @FXML
    public Label numberOfTransactionsLabel;
    @FXML
    public JFXComboBox<String> balanceChartMonthComboBox;
    @FXML
    public JFXComboBox<String> variableExpensesDistributionMonthComboBox;
    @FXML
    public DonutChart variableExpensesDistributionPieChart;
    @FXML
    public SmoothedChart<String, Number> balanceChart;

    private final RestApi restApi = new RestApiImpl();

    private Amount balanceAmount = new Amount();
    private final LocalStorageImpl localStorage = (LocalStorageImpl) LocalStorageImpl.getInstance();
    private final JavaFXFormatter formatter = new JavaFXFormatter(localStorage);
    private CategoryRoot categoryRoot;
    private User user;


    @Override
    public void initialize(URL location, ResourceBundle resources) {
        FinancerController.setInitializationThread(new Thread(() -> {
            FinancerController.getInstance().showLoadingBox();
            categoryRoot = this.localStorage.readObject("categories");
            user = localStorage.readObject("user");

            loadDetailedBalance();
            loadLatestTransactions();
            loadUpcomingFixedTransactions();

            loadBalanceWidget();
            loadVariableExpensesWidget();
            loadNumberOfTransactionsWidget();

            initializeBalanceChart();
            initializeDistributionChart();

            FinancerController.getInstance().hideLoadingBox();
        }));
        FinancerController.getInitializationThread().start();
    }

    private void loadBalanceWidget() {
        final Amount balanceRatio = this.balanceAmount.calcRatio(categoryRoot.getAmount(new ValueDate(LocalDate.now().minusMonths(1))));
        Platform.runLater(() -> {
            this.balanceLabel.setText(formatter.format(this.balanceAmount));
            formatter.format(this.balanceChangeLabel, balanceRatio);
            if (!Double.isNaN(balanceRatio.getAmount()) && Double.isFinite(balanceRatio.getAmount())) {
                this.balanceChangeLabel.setText((balanceRatio.isNegative() ? "" : "+") + this.balanceChangeLabel.getText());
            }
        });
    }

    private void loadVariableExpensesWidget() {
        final Amount variableExpensesAmount = this.categoryRoot.getAmount(CategoryClass.Values.VARIABLE_EXPENSES, new ValueDate());
        final Amount variableExpensesRatio = variableExpensesAmount.calcRatio(this.categoryRoot.getAmount(CategoryClass.Values.VARIABLE_EXPENSES, new ValueDate(LocalDate.now().minusMonths(1))));
        Platform.runLater(() -> {
            this.variableExpensesLabel.setText(formatter.format(variableExpensesAmount));
            formatter.formatChangeLabel(this.variableExpensesChangeLabel, variableExpensesRatio.getAmount());
            if (!Double.isNaN(variableExpensesRatio.getAmount()) && Double.isFinite(variableExpensesRatio.getAmount()))
                this.variableExpensesChangeLabel.setText((variableExpensesRatio.isNegative() ? "" : "+") + this.variableExpensesChangeLabel.getText());
        });
    }

    private void loadNumberOfTransactionsWidget() {
        AtomicInteger numberOfTransactions = new AtomicInteger();
        TreeUtil.traverse(categoryRoot.getCategoriesByClass(CategoryClass.Values.VARIABLE_EXPENSES),
                object -> numberOfTransactions.addAndGet((int) ((Category) object).getTransactions().stream()
                        .filter(transaction -> ((VariableTransaction) transaction).getValueDate().isInSameMonth(new ValueDate()))
                        .count()));
        TreeUtil.traverse(categoryRoot.getCategoriesByClass(CategoryClass.Values.VARIABLE_REVENUE),
                object -> numberOfTransactions.addAndGet((int) ((Category) object).getTransactions().stream()
                        .filter(transaction -> ((VariableTransaction) transaction).getValueDate().isInSameMonth(new ValueDate()))
                        .count()));

        AtomicInteger numberOfTransactionsLastMonth = new AtomicInteger();
        TreeUtil.traverse(categoryRoot.getCategoriesByClass(CategoryClass.Values.VARIABLE_EXPENSES),
                object -> numberOfTransactionsLastMonth.addAndGet((int) ((Category) object).getTransactions().stream()
                        .filter(transaction -> ((VariableTransaction) transaction).getValueDate().isInSameMonth(new ValueDate(LocalDate.now().minusMonths(1))))
                        .count()));
        TreeUtil.traverse(categoryRoot.getCategoriesByClass(CategoryClass.Values.VARIABLE_REVENUE),
                object -> numberOfTransactionsLastMonth.addAndGet((int) ((Category) object).getTransactions().stream()
                        .filter(transaction -> ((VariableTransaction) transaction).getValueDate().isInSameMonth(new ValueDate(LocalDate.now().minusMonths(1))))
                        .count()));

        double numberOfTransactionsRatio = (double) numberOfTransactions.get() / (double) numberOfTransactionsLastMonth.get() * 100 - 100;
        Platform.runLater(() -> {
            this.numberOfTransactionsLabel.setText(Integer.toString(numberOfTransactions.get()));
            formatter.formatChangeLabel(this.numberOfTransactionsChangeLabel, numberOfTransactionsRatio);
            if (!Double.isNaN(numberOfTransactionsRatio) && Double.isFinite(numberOfTransactionsRatio))
                this.numberOfTransactionsChangeLabel.setText((numberOfTransactionsRatio < 0 ? "" : "+")
                        + this.numberOfTransactionsChangeLabel.getText());
        });
    }

    private void loadLatestTransactions() {
        List<VariableTransaction> transactions = new ArrayList<>();
        categoryRoot.traverse(categoryTree -> {
            for (Transaction transaction : ((Category) categoryTree).getTransactions()) {
                if (!transaction.isFixed()) {
                    transactions.add((VariableTransaction) transaction);
                }
            }
        });
        transactions.sort((o1, o2) -> o2.getValueDate().compareTo(o1.getValueDate()));
        this.lastTransactionsGridPane.setVgap(8);
        if (!transactions.isEmpty()) {
            int counter = 0;
            for (VariableTransaction transaction : transactions) {
                // LAST TRANSACTIONS
                if (counter >= 7) {
                    break;
                }
                final int _counter = counter;
                Platform.runLater(() -> this.lastTransactionsGridPane.add(new Label(transaction.getCategory().getName()),
                        0, _counter));

                Label amountLabel = formatter.format(new Label(), transaction.getAmount());
                Platform.runLater(() -> this.lastTransactionsGridPane.add(amountLabel, 1, _counter));
                GridPane.setHalignment(amountLabel, HPos.RIGHT);
                GridPane.setHgrow(amountLabel, Priority.ALWAYS);
                GridPane.setVgrow(amountLabel, Priority.ALWAYS);
                counter++;
            }
        } else {
            Platform.runLater(() -> this.lastTransactionsGridPane.add(new Label(I18N.get("noRecentTransactions")), 0, 0));
        }
    }

    private void loadDetailedBalance() {
        int counter = 0;
        for (CategoryClass categoryClass : CategoryClass.getAll()) {
            final int _counter = counter;
            Platform.runLater(() -> balanceGridPane.add(new Label(I18N.get(categoryClass.getCategoryClass().getName())), 0, _counter));
            final Amount amount = categoryRoot.getAmount(categoryClass.getCategoryClass(), new ValueDate());
            Label baseCategoryLabel = formatter.format(new Label(), amount);
            balanceAmount = balanceAmount.add(amount);
            GridPane.setHalignment(baseCategoryLabel, HPos.RIGHT);
            GridPane.setHgrow(baseCategoryLabel, Priority.ALWAYS);
            GridPane.setVgrow(baseCategoryLabel, Priority.ALWAYS);
            Platform.runLater(() -> balanceGridPane.add(baseCategoryLabel, 1, _counter));
            counter++;
        }

        Label balanceTextLabel = new Label(I18N.get("balance"));
        balanceTextLabel.setId("balance-label");
        Platform.runLater(() -> balanceGridPane.add(balanceTextLabel, 0, 4));
        Label balanceLabel = formatter.format(new Label(), balanceAmount);
        balanceLabel.setId("balance-amount");
        GridPane.setHalignment(balanceLabel, HPos.RIGHT);
        GridPane.setHgrow(balanceLabel, Priority.ALWAYS);
        GridPane.setVgrow(balanceLabel, Priority.ALWAYS);
        Platform.runLater(() -> balanceGridPane.add(balanceLabel, 1, 4));
    }

    private void loadUpcomingFixedTransactions() {
        List<FixedTransaction> transactions = new ArrayList<>();
        categoryRoot.traverse(categoryTree -> {
            for (Transaction transaction : ((Category) categoryTree).getTransactions()) {
                if (transaction.isFixed() && ((FixedTransaction) transaction).isActive() &&
                        (transaction.getAmount(new ValueDate()).getAmount() == 0)) {
                    transactions.add((FixedTransaction) transaction);
                }
            }
        });

        if (!transactions.isEmpty()) {
            int counter = 0;
            for (FixedTransaction transaction : transactions) {
                // LAST TRANSACTIONS
                if (counter >= 5) {
                    break;
                }
                final int _counter = counter;
                if (transaction.getIsVariable()) {
                    Hyperlink link = new Hyperlink(transaction.getCategory().getName());
                    link.setOnAction(event -> addTransactionAmount(transaction));
                    Platform.runLater(() -> this.upcomingFixedTransactionGridPane.add(link, 0, _counter));
                } else {
                    Platform.runLater(() -> this.upcomingFixedTransactionGridPane.add(new Label(transaction.getCategory().getName()),
                            0, _counter));
                }

                Label dayLabel = new Label(formatter.format(LocalDate.now().withDayOfMonth(transaction.getDay())));
                Platform.runLater(() -> this.upcomingFixedTransactionGridPane.add(dayLabel, 1, _counter));
                GridPane.setHalignment(dayLabel, HPos.RIGHT);
                GridPane.setHgrow(dayLabel, Priority.ALWAYS);
                GridPane.setVgrow(dayLabel, Priority.ALWAYS);
                counter++;
            }
        } else {
            Platform.runLater(() -> this.upcomingFixedTransactionGridPane.add(new Label(I18N.get("noUpcomingTransactions")), 0, 0));
        }
    }

    private void addTransactionAmount(FixedTransaction transaction) {
        FixedTransactionAmount transactionAmount = new FixedTransactionAmount()
                .setId(0)
                .setAmount(new Amount())
                .setValueDate(new ValueDate());
        transactionAmount.setFixedTransaction(transaction);
        TransactionAmountDialog dialog = new TransactionAmountDialog(transactionAmount, new ArrayList<>(transaction.getTransactionAmounts()));

        dialog.setOnConfirm(result -> {
            restApi.createTransactionAmount(transaction.getId(), transactionAmount, fixedTransactionAmount -> {
                transaction.getTransactionAmounts().add(transactionAmount);
                localStorage.writeObject("categories", categoryRoot);
                Platform.runLater(() -> upcomingFixedTransactionGridPane.getChildren().clear());
                loadUpcomingFixedTransactions();
            }).execute();
        });
    }

    private void initializeBalanceChart() {
        balanceChart.setChartType(SmoothedChart.ChartType.AREA);
        balanceChart.setAnimated(false);
        this.balanceChartMonthComboBox.getItems().add(I18N.get("lastMonths", 3));
        this.balanceChartMonthComboBox.getItems().add(I18N.get("lastMonths", 6));
        this.balanceChartMonthComboBox.getItems().add(I18N.get("lastMonths", 12));
        this.balanceChartMonthComboBox.valueProperty().addListener(
                (options, oldValue, newValue) -> loadBalanceChartData());
        Platform.runLater(() -> this.balanceChartMonthComboBox.getSelectionModel().select(0));
    }

    private void loadBalanceChartData() {
        balanceChart.getData().clear();
        XYChart.Series<String, Number> data = new XYChart.Series<>();
        data.setName(I18N.get("balance"));
        int numberOfMonths = (int) (1.5 * Math.pow(this.balanceChartMonthComboBox.getSelectionModel().getSelectedIndex(), 2)
                + 1.5 * this.balanceChartMonthComboBox.getSelectionModel().getSelectedIndex() + 3);
        for (int i = numberOfMonths - 1; i >= 0; i--) {
            LocalDate date = LocalDate.now().minusMonths(i);
            XYChart.Data<String, Number> dataSet = new XYChart.Data<>(formatter.format(date), categoryRoot.getAmount(new ValueDate(date)).getAmount());
            Platform.runLater(() -> Tooltip.install(dataSet.getNode(),
                    new Tooltip(dataSet.getXValue() + "\n" +
                            I18N.get("amount") + ": \t" + formatter.format((Double) dataSet.getYValue()))));
            data.getData().add(dataSet);
        }

        balanceChart.getData().add(data);
    }

    private void initializeDistributionChart() {
        this.variableExpensesDistributionMonthComboBox.getItems().add(I18N.get("thisMonth"));
        this.variableExpensesDistributionMonthComboBox.getItems().add(I18N.get("lastMonths", 3));
        this.variableExpensesDistributionMonthComboBox.getItems().add(I18N.get("lastMonths", 6));
        this.variableExpensesDistributionMonthComboBox.getItems().add(I18N.get("lastMonths", 12));
        this.variableExpensesDistributionMonthComboBox.valueProperty().addListener(
                (options, oldValue, newValue) -> loadDistributionChartData());
        Platform.runLater(() -> this.variableExpensesDistributionMonthComboBox.getSelectionModel().select(0));
    }

    private void loadDistributionChartData() {
        ObservableList<PieChart.Data> variableExpensesData = FXCollections.observableArrayList();
        for (Category category : this.categoryRoot.getCategoriesByClass(CategoryClass.Values.VARIABLE_EXPENSES)) {
            Amount amount;
            if (this.variableExpensesDistributionMonthComboBox.getSelectionModel().getSelectedIndex() == 0) {
                amount = category.getAmount(new ValueDate());
            } else {
                int numberOfMonths = (int) (1.5 * Math.pow(this.variableExpensesDistributionMonthComboBox.getSelectionModel().getSelectedIndex(), 2)
                        + 1.5 * this.balanceChartMonthComboBox.getSelectionModel().getSelectedIndex() + 3);
                amount = category.getAmount(new TimeRange(LocalDate.now().minusMonths(numberOfMonths), LocalDate.now()));
            }
            if (amount.isNotNull()) {
                PieChart.Data data = new PieChart.Data(category.getName(), Math.abs(amount.getAmount()));
                Platform.runLater(() -> Tooltip.install(data.getNode(),
                        new Tooltip(formatter.format(category) + "\n" +
                                I18N.get("amount") + ": \t" + formatter.format(data.getPieValue()))));
                variableExpensesData.add(data);
            }
        }
        this.variableExpensesDistributionPieChart.setData(variableExpensesData);
    }
}

File: java/org.financer.client.javafx/src/main/kotlin/org/financer/client/javafx/main/overview/OverviewController.kt
package org.financer.client.javafx.main.overview

import com.jfoenix.controls.JFXComboBox
import javafx.application.Platform
import javafx.beans.value.ObservableValue
import javafx.collections.FXCollections
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.geometry.HPos
import javafx.scene.chart.PieChart
import javafx.scene.chart.XYChart
import javafx.scene.control.Hyperlink
import javafx.scene.control.Label
import javafx.scene.control.Tooltip
import javafx.scene.layout.GridPane
import javafx.scene.layout.Priority
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.financer.client.domain.api.RestApi
import org.financer.client.domain.api.RestApiImpl
import org.financer.client.domain.model.category.Category
import org.financer.client.domain.model.category.CategoryRoot
import org.financer.client.domain.model.transaction.FixedTransaction
import org.financer.client.domain.model.transaction.FixedTransactionAmount
import org.financer.client.domain.model.transaction.Transaction
import org.financer.client.domain.model.transaction.VariableTransaction
import org.financer.client.domain.model.user.User
import org.financer.client.format.I18N.get
import org.financer.client.javafx.components.charts.DonutChart
import org.financer.client.javafx.components.charts.SmoothedChart
import org.financer.client.javafx.format.JavaFXFormatter
import org.financer.client.javafx.local.LocalStorageImpl
import org.financer.client.javafx.main.FinancerController
import org.financer.client.javafx.main.transactions.TransactionAmountDialog
import org.financer.shared.domain.model.value.objects.Amount
import org.financer.shared.domain.model.value.objects.CategoryClass
import org.financer.shared.domain.model.value.objects.TimeRange
import org.financer.shared.domain.model.value.objects.ValueDate
import org.financer.util.collections.Action
import org.financer.util.collections.Tree
import org.financer.util.collections.TreeUtil
import java.net.URL
import java.time.LocalDate
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class OverviewController : Initializable {
    @FXML
    var lastTransactionsGridPane: GridPane? = null

    @FXML
    var balanceGridPane: GridPane? = null

    @FXML
    var upcomingFixedTransactionGridPane: GridPane? = null

    @FXML
    var balanceChangeLabel: Label? = null

    @FXML
    var balanceLabel: Label? = null

    @FXML
    var variableExpensesLabel: Label? = null

    @FXML
    var variableExpensesChangeLabel: Label? = null

    @FXML
    var numberOfTransactionsChangeLabel: Label? = null

    @FXML
    var numberOfTransactionsLabel: Label? = null

    @FXML
    var balanceChartMonthComboBox: JFXComboBox<String>? = null

    @FXML
    var variableExpensesDistributionMonthComboBox: JFXComboBox<String>? = null

    @FXML
    var variableExpensesDistributionPieChart: DonutChart? = null

    @FXML
    var balanceChart: SmoothedChart<String?, Number>? = null
    private val restApi: RestApi = RestApiImpl()
    private var balanceAmount = Amount()
    private val localStorage = LocalStorageImpl.getInstance() as LocalStorageImpl
    private val formatter = JavaFXFormatter(localStorage)
    private lateinit var categoryRoot: CategoryRoot
    private var user: User? = null
    override fun initialize(location: URL, resources: ResourceBundle) {
        CoroutineScope(Dispatchers.Default).launch {
            FinancerController.instance!!.showLoadingBox()
            categoryRoot = localStorage.readObject("categories")
                    ?: throw IllegalStateException("There are no categories stored")
            user = localStorage.readObject("user")
            loadDetailedBalance()
            loadLatestTransactions()
            loadUpcomingFixedTransactions()
            loadBalanceWidget()
            loadVariableExpensesWidget()
            loadNumberOfTransactionsWidget()
            initializeBalanceChart()
            initializeDistributionChart()
            FinancerController.instance!!.hideLoadingBox()
        }
    }

    private fun loadBalanceWidget() {
        val balanceRatio = balanceAmount.calcRatio(categoryRoot!!.getAmount(ValueDate(LocalDate.now().minusMonths(1))))
        Platform.runLater {
            balanceLabel!!.text = formatter.format(balanceAmount)
            formatter.format(balanceChangeLabel!!, balanceRatio)
            if (!java.lang.Double.isNaN(balanceRatio.amount) && java.lang.Double.isFinite(balanceRatio.amount)) {
                balanceChangeLabel!!.text = (if (balanceRatio.isNegative) "" else "+") + balanceChangeLabel!!.text
            }
        }
    }

    private fun loadVariableExpensesWidget() {
        val variableExpensesAmount = categoryRoot!!.getAmount(CategoryClass.Values.VARIABLE_EXPENSES, ValueDate())
        val variableExpensesRatio = variableExpensesAmount.calcRatio(categoryRoot!!.getAmount(CategoryClass.Values.VARIABLE_EXPENSES, ValueDate(LocalDate.now().minusMonths(1))))
        Platform.runLater {
            variableExpensesLabel!!.text = formatter.format(variableExpensesAmount)
            formatter.formatChangeLabel(variableExpensesChangeLabel!!, variableExpensesRatio.amount)
            if (!java.lang.Double.isNaN(variableExpensesRatio.amount) && java.lang.Double.isFinite(variableExpensesRatio.amount)) variableExpensesChangeLabel!!.text = (if (variableExpensesRatio.isNegative) "" else "+") + variableExpensesChangeLabel!!.text
        }
    }

    private fun loadNumberOfTransactionsWidget() {
        val numberOfTransactions = AtomicInteger()
        TreeUtil.traverse(categoryRoot!!.getCategoriesByClass(CategoryClass.Values.VARIABLE_EXPENSES)
        ) { `object`: Tree ->
            numberOfTransactions.addAndGet((`object` as Category).transactions.stream()
                    .filter { transaction: Transaction -> (transaction as VariableTransaction).valueDate!!.isInSameMonth(ValueDate()) }
                    .count().toInt())
        }
        TreeUtil.traverse(categoryRoot!!.getCategoriesByClass(CategoryClass.Values.VARIABLE_REVENUE)
        ) { `object`: Tree ->
            numberOfTransactions.addAndGet((`object` as Category).transactions.stream()
                    .filter { transaction: Transaction -> (transaction as VariableTransaction).valueDate!!.isInSameMonth(ValueDate()) }
                    .count().toInt())
        }
        val numberOfTransactionsLastMonth = AtomicInteger()
        TreeUtil.traverse(categoryRoot!!.getCategoriesByClass(CategoryClass.Values.VARIABLE_EXPENSES)
        ) { `object`: Tree ->
            numberOfTransactionsLastMonth.addAndGet((`object` as Category).transactions.stream()
                    .filter { transaction: Transaction -> (transaction as VariableTransaction).valueDate!!.isInSameMonth(ValueDate(LocalDate.now().minusMonths(1))) }
                    .count().toInt())
        }
        TreeUtil.traverse(categoryRoot!!.getCategoriesByClass(CategoryClass.Values.VARIABLE_REVENUE)
        ) { `object`: Tree ->
            numberOfTransactionsLastMonth.addAndGet((`object` as Category).transactions.stream()
                    .filter { transaction: Transaction -> (transaction as VariableTransaction).valueDate!!.isInSameMonth(ValueDate(LocalDate.now().minusMonths(1))) }
                    .count().toInt())
        }
        val numberOfTransactionsRatio = numberOfTransactions.get().toDouble() / numberOfTransactionsLastMonth.get().toDouble() * 100 - 100
        Platform.runLater {
            numberOfTransactionsLabel!!.text = Integer.toString(numberOfTransactions.get())
            formatter.formatChangeLabel(numberOfTransactionsChangeLabel!!, numberOfTransactionsRatio)
            if (!java.lang.Double.isNaN(numberOfTransactionsRatio) && java.lang.Double.isFinite(numberOfTransactionsRatio)) numberOfTransactionsChangeLabel!!.text = ((if (numberOfTransactionsRatio < 0) "" else "+")
                    + numberOfTransactionsChangeLabel!!.text)
        }
    }

    private fun loadLatestTransactions() {
        val transactions: MutableList<VariableTransaction> = ArrayList()
        categoryRoot!!.traverse { categoryTree: Tree ->
            for (transaction in (categoryTree as Category).transactions) {
                if (!transaction.isFixed) {
                    transactions.add(transaction as VariableTransaction)
                }
            }
        }
        transactions.sortWith(Comparator { o1: VariableTransaction, o2: VariableTransaction -> o2.valueDate!!.compareTo(o1.valueDate) })
        lastTransactionsGridPane!!.vgap = 8.0
        if (!transactions.isEmpty()) {
            var counter = 0
            for (transaction in transactions) {
                // LAST TRANSACTIONS
                if (counter >= 7) {
                    break
                }
                val _counter = counter
                Platform.runLater {
                    lastTransactionsGridPane!!.add(Label(transaction.category!!.name),
                            0, _counter)
                }
                val amountLabel = formatter.format(Label(), transaction.amount)
                Platform.runLater { lastTransactionsGridPane!!.add(amountLabel, 1, _counter) }
                GridPane.setHalignment(amountLabel, HPos.RIGHT)
                GridPane.setHgrow(amountLabel, Priority.ALWAYS)
                GridPane.setVgrow(amountLabel, Priority.ALWAYS)
                counter++
            }
        } else {
            Platform.runLater { lastTransactionsGridPane!!.add(Label(get("noRecentTransactions")), 0, 0) }
        }
    }

    private fun loadDetailedBalance() {
        var counter = 0
        for (categoryClass in CategoryClass.getAll()) {
            val _counter = counter
            Platform.runLater { balanceGridPane!!.add(Label(get(categoryClass.categoryClass.getName())), 0, _counter) }
            val amount = categoryRoot!!.getAmount(categoryClass.categoryClass, ValueDate())
            val baseCategoryLabel = formatter.format(Label(), amount)
            balanceAmount = balanceAmount.add(amount)
            GridPane.setHalignment(baseCategoryLabel, HPos.RIGHT)
            GridPane.setHgrow(baseCategoryLabel, Priority.ALWAYS)
            GridPane.setVgrow(baseCategoryLabel, Priority.ALWAYS)
            Platform.runLater { balanceGridPane!!.add(baseCategoryLabel, 1, _counter) }
            counter++
        }
        val balanceTextLabel = Label(get("balance"))
        balanceTextLabel.id = "balance-label"
        Platform.runLater { balanceGridPane!!.add(balanceTextLabel, 0, 4) }
        val balanceLabel = formatter.format(Label(), balanceAmount)
        balanceLabel.id = "balance-amount"
        GridPane.setHalignment(balanceLabel, HPos.RIGHT)
        GridPane.setHgrow(balanceLabel, Priority.ALWAYS)
        GridPane.setVgrow(balanceLabel, Priority.ALWAYS)
        Platform.runLater { balanceGridPane!!.add(balanceLabel, 1, 4) }
    }

    private fun loadUpcomingFixedTransactions() {
        val transactions: MutableList<FixedTransaction> = ArrayList()
        categoryRoot!!.traverse { categoryTree: Tree ->
            for (transaction in (categoryTree as Category).transactions) {
                if (transaction.isFixed && (transaction as FixedTransaction).isActive &&
                        transaction.getAmount(ValueDate()).amount == 0.0) {
                    transactions.add(transaction)
                }
            }
        }
        if (!transactions.isEmpty()) {
            var counter = 0
            for (transaction in transactions) {
                // LAST TRANSACTIONS
                if (counter >= 5) {
                    break
                }
                val _counter = counter
                if (transaction.isVariable) {
                    val link = Hyperlink(transaction.category!!.name)
                    link.onAction = EventHandler { event: ActionEvent? -> addTransactionAmount(transaction) }
                    Platform.runLater { upcomingFixedTransactionGridPane!!.add(link, 0, _counter) }
                } else {
                    Platform.runLater {
                        upcomingFixedTransactionGridPane!!.add(Label(transaction.category!!.name),
                                0, _counter)
                    }
                }
                val dayLabel = Label(formatter.format(LocalDate.now().withDayOfMonth(transaction.day)))
                Platform.runLater { upcomingFixedTransactionGridPane!!.add(dayLabel, 1, _counter) }
                GridPane.setHalignment(dayLabel, HPos.RIGHT)
                GridPane.setHgrow(dayLabel, Priority.ALWAYS)
                GridPane.setVgrow(dayLabel, Priority.ALWAYS)
                counter++
            }
        } else {
            Platform.runLater { upcomingFixedTransactionGridPane!!.add(Label(get("noUpcomingTransactions")), 0, 0) }
        }
    }

    private fun addTransactionAmount(transaction: FixedTransaction) {
        val transactionAmount = FixedTransactionAmount()
                .setId(0)
                .setAmount(Amount())
                .setValueDate(ValueDate())
        transactionAmount.setFixedTransaction(transaction)
        val dialog = TransactionAmountDialog(transactionAmount, ArrayList(transaction.getTransactionAmounts()))
        dialog.setOnConfirm(Action {
            CoroutineScope(Dispatchers.IO).launch {
                val updatedTransactionAmount = restApi.createTransactionAmount(transaction.id, transactionAmount)
                if (updatedTransactionAmount != null) {
                    transaction.getTransactionAmounts().add(updatedTransactionAmount)
                    localStorage.writeObject("categories", categoryRoot)
                    Platform.runLater { upcomingFixedTransactionGridPane!!.children.clear() }
                    loadUpcomingFixedTransactions()
                }
            }
        })
    }

    private fun initializeBalanceChart() {
        balanceChart!!.setChartType(SmoothedChart.ChartType.AREA)
        balanceChart!!.animated = false
        balanceChartMonthComboBox!!.items.add(get("lastMonths", 3))
        balanceChartMonthComboBox!!.items.add(get("lastMonths", 6))
        balanceChartMonthComboBox!!.items.add(get("lastMonths", 12))
        balanceChartMonthComboBox!!.valueProperty().addListener { options: ObservableValue<out String>?, oldValue: String?, newValue: String? -> loadBalanceChartData() }
        Platform.runLater { balanceChartMonthComboBox!!.selectionModel.select(0) }
    }

    private fun loadBalanceChartData() {
        balanceChart!!.data.clear()
        val data = XYChart.Series<String?, Number>()
        data.name = get("balance")
        val numberOfMonths = (1.5 * Math.pow(balanceChartMonthComboBox!!.selectionModel.selectedIndex.toDouble(), 2.0) + 1.5 * balanceChartMonthComboBox!!.selectionModel.selectedIndex + 3).toInt()
        for (i in numberOfMonths - 1 downTo 0) {
            val date = LocalDate.now().minusMonths(i.toLong())
            val dataSet = XYChart.Data<String?, Number>(formatter.format(date), categoryRoot!!.getAmount(ValueDate(date)).amount)
            Platform.runLater {
                Tooltip.install(dataSet.node,
                        Tooltip("""
    ${dataSet.xValue.toString()}
    ${get("amount")}: 	${formatter.format(dataSet.yValue as Double)}
    """.trimIndent()))
            }
            data.data.add(dataSet)
        }
        balanceChart!!.data.add(data)
    }

    private fun initializeDistributionChart() {
        variableExpensesDistributionMonthComboBox!!.items.add(get("thisMonth"))
        variableExpensesDistributionMonthComboBox!!.items.add(get("lastMonths", 3))
        variableExpensesDistributionMonthComboBox!!.items.add(get("lastMonths", 6))
        variableExpensesDistributionMonthComboBox!!.items.add(get("lastMonths", 12))
        variableExpensesDistributionMonthComboBox!!.valueProperty().addListener { options: ObservableValue<out String>?, oldValue: String?, newValue: String? -> loadDistributionChartData() }
        Platform.runLater { variableExpensesDistributionMonthComboBox!!.selectionModel.select(0) }
    }

    private fun loadDistributionChartData() {
        val variableExpensesData = FXCollections.observableArrayList<PieChart.Data>()
        for (category in categoryRoot!!.getCategoriesByClass(CategoryClass.Values.VARIABLE_EXPENSES)) {
            var amount: Amount
            amount = if (variableExpensesDistributionMonthComboBox!!.selectionModel.selectedIndex == 0) {
                category.getAmount(ValueDate())
            } else {
                val numberOfMonths = (1.5 * Math.pow(variableExpensesDistributionMonthComboBox!!.selectionModel.selectedIndex.toDouble(), 2.0) + 1.5 * balanceChartMonthComboBox!!.selectionModel.selectedIndex + 3).toInt()
                category.getAmount(TimeRange(LocalDate.now().minusMonths(numberOfMonths.toLong()), LocalDate.now()))
            }
            if (amount.isNotNull) {
                val data = PieChart.Data(category.name, Math.abs(amount.amount))
                Platform.runLater {
                    Tooltip.install(data.node,
                            Tooltip("""
    ${formatter.format(category)}
    ${get("amount")}: 	${formatter.format(data.pieValue)}
    """.trimIndent()))
                }
                variableExpensesData.add(data)
            }
        }
        variableExpensesDistributionPieChart!!.data = variableExpensesData
    }
}